testing/asserts.ts

```diff
@@ -135,7 +136,14 @@ function constructorsEqual(a: object, b: object) {
     !a.constructor && b.constructor === Object;
 }
 
+type IsTruthy<T> = T extends Falsy ? false : true;
+type IsFalsy<T> = T extends Falsy ? true : false;
+
 /** Make an assertion, error will be thrown if `expr` does not have truthy value. */
+export function assert<const T>(
+  expr: T,
+  msg?: string,
+): IsTruthy<T> extends true ? void : never;
 export function assert(expr: unknown, msg = ""): asserts expr {
```

Comment:
This comment applies to this function and the others on which overload signatures have been added.

Adding non-assertion [overload signatures](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) (without `assert` predicates) to [assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions) is an anti-pattern because the purpose of assertion functions in the type system is for narrowing of values' types at runtime.

Here's an example to demonstrate what I mean:

[TS Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&removeComments=true&target=99&jsx=4&exactOptionalPropertyTypes=true&inlineSourceMap=true&inlineSources=true&isolatedModules=true&noImplicitOverride=true#code/CYUwxgNghgTiAEEQBd4CMBc8AM8A+8AjANwBQpA3vAPTXwCCAdsgJYC0ADlMsiDI1gAiAexABnRgHJUjWDGEB3eMgCeHBMIBm8AG5QIAVwRRNvGPANiQpeMrUIAYvrEr4AXnibnCArl+N8eAAiIMDGAwgIQINGUE0WRhBgMltVdXgASTEAFRgDZAALFQAebIA+d3hs+BAAD15YsXgnCBd4AH5Pb3gsZDyQMhtPGLBWYQCoMSsYZGKwcbFUcoAKOo4YLGyASiws3Pyi0oq6huAmvqMO3WEWYB74RJ0+FOHGUZZx+Enp5FXa9awMQA1oxFIwdl8pnxkE01uYKABfci2WjBNCheYRO5oBC4YTmQjwdYfczIYTKArGKEzD4CIbfaHLNBbF62Nm0AB67WRNDoQXR8DEBWEWPg4wgrhxRC+qEKLCaHBuzAwQzQpE53KRlF58AAcnJFE1JjV-uBeMAhpoRmMJtTfnDAYwQWCIQyZrD-vCtbY3b9may2byuTy1RrSEigA)

```ts
declare let b: 0 | 1;

{ // Anti-pattern: Doesn't narrow type of value after use
  type Falsy = false | 0 | 0n | "" | null | undefined;
  type IsTruthy<T> = T extends Falsy ? false : true;

  function assert<const T>(expr: T): IsTruthy<T> extends true ? void : never;
  function assert(expr: unknown): asserts expr {}

  // "b" could be 0 or 1 prior to the assertion:
  assert(b);
       //^? let b: 0 | 1

  // "b" should only be 1 at this point:
  b
//^? let b: 0 | 1
}

{ // Narrows as expected
  function assert(expr: unknown): asserts expr {}

  assert(b);
       //^? let b: 0 | 1

  b
//^? let b: 1
}

```

Reply:
@scarf005 I also noticed that you linked to my comment in your PR message. Please don't take that or this critique personally, but do take them seriously: correct typings can be complicated — I just want to help ensure that if this kind of feature is implemented, it's done with thorough consideration, testing, and review.